llama/llama.go

```diff
@@ -0,0 +1,512 @@
+
+	err := -1
+	if loraAdapter != nil {
+		err = int(C.llama_lora_adapter_set(context.c, loraAdapter, C.float(scale)))
+	}
+	if err != 0 {
+		return errors.New("error applying lora from file")
+	}
+
+	return nil
+}
+
+type Batch struct {
+	c         C.struct_llama_batch
+	batchSize int
+	embedSize int
+}
+
+// Creates a new batch for either word tokens if embed is 0 or
+// image embeddings if embed is specified. Batches cannot contain
+// both types at the same time
+func NewBatch(nTokens int, embed int, maxSeq int) *Batch {
+	return &Batch{
+		c:         C.llama_batch_init(C.int(nTokens), C.int(embed), C.int(maxSeq)),
+		batchSize: nTokens,
+		embedSize: embed,
+	}
+}
+
+func (b *Batch) NumTokens() int {
+	return int(b.c.n_tokens)
+}
+
+func (b *Batch) IsEmbedding() bool {
+	return b.embedSize != 0
+}
+
+// Add adds either a token or an image embedding to the batch depending on the type
+// when the batch was initialized. The other argument will be ignored. Adds to the
+// batch with the given position for the given sequence ids, and optionally instructs
+// to include logits.
+func (b *Batch) Add(token int, embed []float32, pos int, seqIds []int, logits bool) {
+	if !b.IsEmbedding() {
+		unsafe.Slice(b.c.token, b.batchSize)[b.c.n_tokens] = C.llama_token(token)
+	} else {
+		copy(unsafe.Slice((*float32)(b.c.embd), b.batchSize*b.embedSize)[int(b.c.n_tokens)*b.embedSize:], embed)
+	}
+	unsafe.Slice(b.c.pos, b.batchSize)[b.c.n_tokens] = C.llama_pos(pos)
+	unsafe.Slice(b.c.n_seq_id, b.batchSize)[b.c.n_tokens] = C.int(len(seqIds))
+
+	for i, s := range seqIds {
+		unsafe.Slice((unsafe.Slice(b.c.seq_id, b.batchSize)[b.c.n_tokens]), C.int(len(seqIds)))[i] = C.int32_t(s)
+	}
+
+	if logits {
+		unsafe.Slice(b.c.logits, b.batchSize)[b.c.n_tokens] = 1
+	}
+
+	b.c.n_tokens += 1
+}
+
+func (b *Batch) Clear() {
+	b.c.n_tokens = 0
+}
+
+func (b *Batch) Free() {
+	b.batchSize = 0
+	C.llama_batch_free(b.c)
+}
+
+type Model struct {
+	c *C.struct_llama_model
+}
+
+func (m *Model) TokenToPiece(token int) string {
+	tokenLen := 12
+	buf := make([]byte, tokenLen)
+	tokenLen = int(C.llama_token_to_piece(
+		m.c,
+		C.int32_t(token),
+		(*C.char)(unsafe.Pointer(&buf[0])),
+		C.int32_t(tokenLen),
+		C.int32_t(0),
+		C.bool(true),
+	))
+	if tokenLen < 0 {
+		tokenLen = -tokenLen
+
+		buf = make([]byte, tokenLen)
+		C.llama_token_to_piece(
+			m.c,
+			C.int32_t(token),
+			(*C.char)(unsafe.Pointer(&buf[0])),
+			C.int32_t(tokenLen),
+			C.int32_t(0),
+			C.bool(true),
+		)
+	}
+	return strings.TrimRight(string(buf), "\x00")
+}
+
+func (m *Model) Tokenize(text string, addSpecial bool, parseSpecial bool) ([]int, error) {
+	maxTokens := len(text) + 2
+	cTokens := make([]C.llama_token, maxTokens)
+	cText := C.CString(text)
+	defer C.free(unsafe.Pointer(cText))
+
+	result := C.llama_tokenize(
+		m.c,
+		cText,
+		C.int32_t(len(text)),
+		&cTokens[0],
+		C.int32_t(maxTokens),
+		C.bool(addSpecial),
+		C.bool(parseSpecial),
+	)
+
+	if result < 0 {
```

Comment:
Sure. For reference, the reason is if the allocated buffer is too small then it will tell you the size that you need so you can realloc and call again.